app/plugin_test.go

```diff
@@ -592,6 +592,180 @@ func TestPluginSync(t *testing.T) {
 			require.Nil(t, err)
 			require.Len(t, pluginStatus, 1)
 			require.Equal(t, pluginStatus[0].PluginId, "testplugin")
+
+			appErr = th.App.RemovePlugin("testplugin")
+			checkNoError(t, appErr)
+
+			appErr = th.App.RemoveFile(th.App.getBundleStorePath("testplugin.tar.gz"))
+			checkNoError(t, appErr)
+			appErr = th.App.RemoveFile(fmt.Sprintf("%s.sig", th.App.getBundleStorePath("testplugin")))
+			checkNoError(t, appErr)
 		})
 	}
 }
+
+func TestCanaryPlugins(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+	os.MkdirAll("./test-plugins", os.ModePerm)
+	defer os.RemoveAll("./test-plugins")
+
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.PluginSettings.Enable = true
+		*cfg.PluginSettings.Directory = "./test-plugins"
+		*cfg.PluginSettings.ClientDirectory = "./test-client-plugins"
+		*cfg.PluginSettings.RequirePluginSignature = false
+		*cfg.PluginSettings.CanaryTag = "daily"
+	})
+
+	env, err := plugin.NewEnvironment(th.App.NewPluginAPI, "./test-plugins", "./test-client-plugins", th.App.Log)
+	require.NoError(t, err)
+	th.App.SetPluginsEnvironment(env)
+	// New bundle in the file store case
+	path, _ := fileutils.FindDir("tests")
+	fileReader, err := os.Open(filepath.Join(path, "testplugin.tar.gz"))
+	require.NoError(t, err)
+	defer fileReader.Close()
+
+	_, appErr := th.App.WriteFile(fileReader, th.App.getBundleStorePath("testplugin"))
+	checkNoError(t, appErr)
+
+	appErr = th.App.SyncPlugins()
+	checkNoError(t, appErr)
+
+	// Check if installed
+	pluginStatus, err := env.Statuses()
+	require.Nil(t, err)
+	require.Len(t, pluginStatus, 1)
+	require.Equal(t, "testplugin", pluginStatus[0].PluginId)
+
+	// No canary tag
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.PluginSettings.CanaryTag = ""
+	})
+	fileReader, err = os.Open(filepath.Join(path, "testplugin_v2.tar.gz"))
+	require.NoError(t, err)
+	defer fileReader.Close()
+
+	_, appErr = th.App.WriteFile(fileReader, th.App.getBundleStorePath("testplugin_v2"))
+	checkNoError(t, appErr)
+
+	appErr = th.App.SyncPlugins()
+	checkNoError(t, appErr)
+	// should be installed plugin with no canary flag
+	pluginStatus, err = env.Statuses()
+	require.Nil(t, err)
+	require.Len(t, pluginStatus, 1)
```

Comment:
I'm guessing `testplugin_v2` hast the same plugin id as `testplugin`?

Reply:
No they are different. `testplugin` is installed on daily `testplugin_v2` is installed with no canary tag. So when server has no canary tag it shows only one plugin

Reply:
Ah, I see! I think it may not be obvious in the tests how the test files are being setup. I wonder if we can clarify what `WriteFile` is actually recording to the file store, i.e. clearly summarizing that we end up with a `testplugin.tar.gz+daily` and `testplugin2.tar.gz` in this case.

Reply:
@lieut-data added some comments to the test.